source/bluetooth/MicroBitUtilityService.cpp

```diff
@@ -0,0 +1,385 @@
+                    break;
+            }
+            break;
+    }
+}
+
+
+/**
+ * Listen for or ignore  events for processing requests
+ * @param yes true to listen, otherwise ignore
+ */
+void MicroBitUtilityService::listen( bool yes)
+{
+    if ( yes)
+        messageBus.listen( MICROBIT_ID_UTILITY, MICROBIT_EVT_ANY, this, &MicroBitUtilityService::onEvent);
+    else
+        messageBus.ignore( MICROBIT_ID_UTILITY, MICROBIT_EVT_ANY, this, &MicroBitUtilityService::onEvent);
+}
+
+
+/**
+ * Send a reply packet
+ * @param data Data to send
+ * @param length Length if the data
+ * @return DEVICE_OK if sent
+ */
+int MicroBitUtilityService::sendReply( const void *data, uint16_t length)
+{
+    bool ok = notifyChrValue( mbbs_cIdxCTRL, (const uint8_t *) data, length);
+    if ( !ok)
+        return DEVICE_BUSY;
+    workspace->onReplySent( data);
+    return DEVICE_OK;
+}
+
+
+/**
+ * Process the current request. This may block.
+ * @return DEVICE_OK if finished
+ */
+int MicroBitUtilityService::processRequest()
+{
+    int result = DEVICE_OK;
+    
+    if ( workspace && !workspace->lock && workspace->request.type != requestTypeNone)
+    {
+        // Block changes to the request while processing it
+        workspace->lock = true;
+        
+        switch ( workspace->request.type)
+        {
+            case requestTypeLogLength:
+                result = processLogLength();
+                break;
+            case requestTypeLogRead:
+                result = processLogRead();
+                break;
+            default:
+                break;
+        }
+        
+        // Check if finished processing the current request
+        if ( result == DEVICE_OK)
+            workspace->init();
+
+        workspace->lock = false;
+    }
+    
+    if ( workspace && workspace->request.type != requestTypeNone)
+        MicroBitEvent evt( MICROBIT_ID_UTILITY, MICROBIT_ID_UTILITY_PROCESS);
+    return result;
+}
+
+
+/**
+ * Process request typeLogLength
+ * @return DEVICE_OK if finished
+ */
+int MicroBitUtilityService::processLogLength()
+{
+    if ( workspace->replyState == replyStateClear)
+    {
+        requestLog_t *request = (requestLog_t *) &workspace->request;
+        uint32_t length = log.getDataLength( (DataFormat) request->format);
+        workspace->setReply( length);
+    }
+    return sendReply( &workspace->reply, offsetof(reply_t, data) + workspace->replyLength);
+}
+
+
+/**
+ * Process request typeLogRead
+ * @return DEVICE_OK if finished
+ */
+int MicroBitUtilityService::processLogRead()
+{
+    requestLogRead_t *request = (requestLogRead_t *) &workspace->request;
+    
+    while ( request->batchlen)
+    {
+        if ( workspace->replyState == replyStateClear)
+        {
+            int block = min( request->batchlen, sizeof( reply_t) - offsetof( reply_t, data));
+            int result = log.readData(workspace->reply.data, request->index, block, (DataFormat) request->format, request->length);
```

Comment:
Does this mean we're very likely to see a sequence of 20 byte read operations from the log @martinwork? That might add quite a bit of latency, compared to one larger read operation?

Reply:
Won't most of the 20 byte reads be from uBit,log's cache? I suppose there's a fair bit of code for each read to arrive at the memcpy.

Reply:
yes, you're quite right! I forgot we applied the block cache from the file system API into the log reads... All good there then.

I'm not worried about the cache lookup/memcpy cost. It will be insignificant compared to all the BLE/I2C latency...